apps/src/templates/AchievementDialog.jsx

```diff
@@ -0,0 +1,198 @@
+    lineHeight: 1.3,
+    margin: '20px auto',
+  },
+  footer: {
+    position: 'absolute',
+    bottom: 10,
+    right: 20,
+    left: 20,
+  },
+  buttonPrimary: {
+    float: 'right',
+    background: color.orange,
+    color: color.white,
+    border: '1px solid #b07202',
+    borderRadius: 3,
+    boxShadow: 'inset 0 1px 0 0 rgba(255, 255, 255, 0.63)',
+    fontSize: 14,
+    padding: '8px 20px',
+  },
+  buttonSecondary: {
+    background: '#eee',
+    color: '#5b6770',
+    border: '1px solid #c5c5c5',
+  },
+};
+
+const AchievementDialog = Radium(React.createClass({
+  propTypes: {
+    puzzleNumber: React.PropTypes.number,
+    idealBlocks: React.PropTypes.number,
+    actualBlocks: React.PropTypes.number,
+    hintsUsed: React.PropTypes.number,
+    assetUrl: React.PropTypes.func,
+    onContinue: React.PropTypes.func,
+  },
+
+  getInitialState() {
+    return {isOpen: true};
+  },
+
+  handleClose(nextPuzzle = true) {
+    this.setState({isOpen: false});
+    nextPuzzle && this.props.onContinue();
+  },
+
+  getIcon(flag) {
+    return (
+      <span
+        style={[
+          styles.achievement.icon,
+          !flag && styles.achievement.inactive
+        ]}
+      >
+        {flag ?
+          <i className="fa fa-check-square-o" style={styles.absolute}/> :
+          [
+            <i
+              key="a"
+              className="fa fa-square-o"
+              style={styles.absolute}
+            />,
+            <i
+              key="b"
+              className="fa fa-times"
+              style={[styles.absolute, {fontSize: 24, left: 2, top: 10}]}
+            />
+          ]
+        }
+      </span>
+    );
+  },
+
+  render() {
+    const params = {
+      puzzleNumber: this.props.puzzleNumber,
+      numBlocks: this.props.idealBlocks
+    };
+    const blockDelta = this.props.actualBlocks - this.props.idealBlocks;
+    const tooManyBlocks = blockDelta > 0;
+    const footerMessage = locale[tooManyBlocks ? 'numBlocksNeeded' : 'nextLevel'](params);
+
+    const tooManyHints = this.props.hintsUsed > 0;
+
+    return (
+      <BaseDialog
+        useUpdatedStyles
+        isOpen={this.state.isOpen}
+        handleClose={this.handleClose.bind(this, !tooManyBlocks)}
+        assetUrl={this.props.assetUrl}
+      >
+        <div style={styles.checkmarks}>
+          <p style={styles.achievement.row}>
+            {this.getIcon(true)}
+            <span style={styles.achievement.text}>
+              {locale.puzzleCompleted()}
+            </span>
+          </p>
+          <p style={styles.achievement.row}>
+            {this.getIcon(!tooManyBlocks)}
+            <span
+              style={[
+                styles.achievement.text,
+                tooManyBlocks && styles.achievement.inactive
+              ]}
+            >
+              {tooManyBlocks ? locale.usingTooManyBlocks(params) :
+                (blockDelta === 0 ? locale.exactNumberOfBlocks(params) :
+                  locale.fewerNumberOfBlocks(params)
```

Comment:
Good call.  I tried extracting these into functions in https://github.com/code-dot-org/code-dot-org/pull/13175/commits/4880f8eea116ef1159b2731fcdb365e339f61139 — thoughts?